File: 0338-counting-bits/0338-counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]: 
        def bit_count(a):
            cnt=0
            while a:
                cnt+= (a&1)
                a >>=1
            return cnt
        bf=[0]
        for i in range(1,n+1):
            bf.append(bit_count(i))
        return bf
        
        # dp = [0] * (n + 1)
        # offset = 1
        # for i in range(1, n + 1):
        #     if offset * 2 == i:
        #         offset = i
        #     dp[i] = 1 + dp[i - offset]
        # return dp

File: 0338-counting-bits/0338-counting-bits.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> ans;
        for(auto i=0;i<=n;i++){
            int cnt=0;
            int t=i;
            while(t != 0){
                if(t&1)cnt+=1;
                t= t>>1;
            }
            ans.push_back(cnt);
        }
        return ans;
    }
};